go-controller/pkg/ovn/loadbalancer.go

```diff
@@ -82,45 +84,179 @@ func (ovn *Controller) getLoadBalancerVIPS(
 
-	// With service_ip:port as a VIP, create an entry in 'load_balancer'
-	// key is of the form "IP:port" (with quotes around)
-	key := fmt.Sprintf(`"%s"`, util.JoinHostPortInt32(serviceIP, port))
+	var commaSeparatedEndpoints string
 
-	if len(ips) == 0 {
-		_, _, err := util.RunOVNNbctl("remove", "load_balancer", lb,
-			"vips", key)
-		return err
+	if len(ips) > 0 {
+		for i, ep := range ips {
+			comma := ","
+			if i == 0 {
+				comma = ""
+			}
+			commaSeparatedEndpoints += fmt.Sprintf("%s%s", comma, util.JoinHostPortInt32(ep, targetPort))
+		}
+
+		// ensure the ACL is removed if it exists
+		ovn.deleteLoadBalancerRejectACL(lb, util.JoinHostPortInt32(serviceIP, port))
 	}
 
-	var commaSeparatedEndpoints string
-	for i, ep := range ips {
-		comma := ","
-		if i == 0 {
-			comma = ""
+	return ovn.configureLoadBalancer(lb, serviceIP, port, commaSeparatedEndpoints)
+}
+
+func (ovn *Controller) getLogicalSwitchesForLoadBalancer(lb string) ([]string, error) {
+	out, _, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=_uuid", "find",
+		"logical_switch", fmt.Sprintf("load_balancer{>=}%s", lb))
+	if err != nil {
+		return nil, err
+	}
+	if len(strings.Fields(out)) > 0 {
+		return strings.Fields(out), nil
+	}
+	// if load balancer was not on a switch, then it may be on a router
+	out, _, err = util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=name", "find",
+		"logical_router", fmt.Sprintf("load_balancer{>=}%s", lb))
+	if err != nil {
+		return nil, err
+	}
+	if len(out) == 0 {
+		return nil, nil
+	}
+	// if this is a GR we know the corresponding join and external switches, otherwise this is an unhandled
+	// case
+	if strings.HasPrefix(out, util.GWRouterPrefix) {
+		routerName := strings.TrimPrefix(out, util.GWRouterPrefix)
+		return []string{util.JoinSwitchPrefix + routerName, util.ExternalSwitchPrefix + routerName}, nil
+	}
+	return nil, fmt.Errorf("router detected with load balancer that is not a GR")
+}
+
+func (ovn *Controller) createLoadBalancerRejectACL(lb string, serviceIP string, port int32, proto kapi.Protocol) (string, error) {
+
+	switches, err := ovn.getLogicalSwitchesForLoadBalancer(lb)
+	if err != nil {
+		return "", fmt.Errorf("error finding logical switch that contains load balancer %s: %v", lb, err)
+	}
+
+	if len(switches) == 0 {
+		klog.Infof("Ignoring creating reject ACL for load balancer %s. It has no logical switches", lb)
+		return "", nil
+	}
+
+	ip := net.ParseIP(serviceIP)
+	if ip == nil {
+		klog.Errorf("Cannot parse IP address %s", serviceIP)
+		return "", fmt.Errorf("cannot create reject ACL, invalid cluster IP: %s", serviceIP)
+	}
+	var aclMatch string
+	var l3Prefix string
+	if ip.To4() != nil {
+		l3Prefix = "ip4"
+	} else {
+		l3Prefix = "ip6"
+	}
+	vip := util.JoinHostPortInt32(serviceIP, port)
+	if eps, ok := ovn.GetServiceLBEndpoints(lb, vip); ok && len(eps) > 0 {
+		klog.Warningf("ACL being applied to Load Balancer with Endpoints: %s, %s", lb, vip)
```

Comment:
@trozet isn't this a race condition here? If there are endpoints present for a given service, shouldn't we just return from here?

Reply:
In certain cases it is intentional. Like here I apply the acl just before deleting the endpoints:
https://github.com/ovn-org/ovn-kubernetes/pull/1096/files#diff-e6c06e0564fcf466a729f740e8722eedR284

I just wanted to add this warning message in case in the future there is a situation where an ACL was applied to something that still had endpoints and it was unintentional. Then we would have something in the logs to point to.